File: bench/generate_mesh.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <fstream>
#include <map>
#include <algorithm>
#include <string>

struct Vector3 {
    double x, y, z;
    Vector3(double x=0, double y=0, double z=0) : x(x), y(y), z(z) {}
    Vector3 operator+(const Vector3& v) const { return Vector3(x+v.x, y+v.y, z+v.z); }
    Vector3 operator/(double d) const { return Vector3(x/d, y/d, z/d); }
    double length() const { return std::sqrt(x*x + y*y + z*z); }
    void normalize() {
        double l = length();
        if (l > 0) { x /= l; y /= l; z /= l; }
    }
};

struct Face {
    int v1, v2, v3;
};

int get_middle_point(int p1_idx, int p2_idx, std::vector<Vector3>& verts, std::map<std::pair<int, int>, int>& cache) {
    int smaller = std::min(p1_idx, p2_idx);
    int greater = std::max(p1_idx, p2_idx);
    auto key = std::make_pair(smaller, greater);

    if (cache.count(key)) return cache[key];

    Vector3 v1 = verts[p1_idx];
    Vector3 v2 = verts[p2_idx];
    Vector3 middle = (v1 + v2) / 2.0;
    middle.normalize();

    verts.push_back(middle);
    int index = verts.size() - 1;
    cache[key] = index;
    return index;
}

void generate_mesh(const std::string& type, int subdivisions, std::vector<Vector3>& verts, std::vector<Face>& faces) {
    if (type == "octahedron") {
        verts = {
            { 1, 0, 0}, {-1, 0, 0}, { 0, 1, 0}, { 0,-1, 0}, { 0, 0, 1}, { 0, 0,-1}
        };
        faces = {
            {4, 0, 2}, {4, 2, 1}, {4, 1, 3}, {4, 3, 0},
            {5, 2, 0}, {5, 1, 2}, {5, 3, 1}, {5, 0, 3}
        };
    } else { // icosahedron (sphere)
        double f = (1.0 + std::sqrt(5.0)) / 2.0;
        verts = {
            {-1,  f,  0}, { 1,  f,  0}, {-1, -f,  0}, { 1, -f,  0},
            { 0, -1,  f}, { 0,  1,  f}, { 0, -1, -f}, { 0,  1, -f},
            { f,  0, -1}, { f,  0,  1}, {-f,  0, -1}, {-f,  0,  1}
        };
        faces = {
            {0, 11, 5}, {0, 5, 1}, {0, 1, 7}, {0, 7, 10}, {0, 10, 11},
            {1, 5, 9}, {5, 11, 4}, {11, 10, 2}, {10, 7, 6}, {7, 1, 8},
            {3, 9, 4}, {3, 4, 2}, {3, 2, 6}, {3, 6, 8}, {3, 8, 9},
            {4, 9, 5}, {2, 4, 11}, {6, 2, 10}, {8, 6, 7}, {9, 8, 1}
        };
    }

    for (auto& v : verts) v.normalize();

    for (int i = 0; i < subdivisions; ++i) {
        std::vector<Face> faces_subdiv;
        std::map<std::pair<int, int>, int> cache;
        for (const auto& tri : faces) {
            int a = get_middle_point(tri.v1, tri.v2, verts, cache);
            int b = get_middle_point(tri.v2, tri.v3, verts, cache);
            int c = get_middle_point(tri.v3, tri.v1, verts, cache);
            faces_subdiv.push_back({tri.v1, a, c});
            faces_subdiv.push_back({tri.v2, b, a});
            faces_subdiv.push_back({tri.v3, c, b});
            faces_subdiv.push_back({a, b, c});
        }
        faces = std::move(faces_subdiv);
    }
}

void save_off(const std::string& filename, const std::vector<Vector3>& verts, const std::vector<Face>& faces) {
    std::ofstream f(filename);
    f << "OFF\n";
    f << verts.size() << " " << faces.size() << " 0\n";
    for (const auto& v : verts) {
        f << v.x << " " << v.y << " " << v.z << "\n";
    }
    for (const auto& tri : faces) {
        f << "3 " << tri.v1 << " " << tri.v2 << " " << tri.v3 << "\n";
    }
}

int main(int argc, char** argv) {
    if (argc < 4) {
        std::cout << "Usage: ./generate_mesh <octahedron|sphere> <subdivisions> <output.off>\n";
        return 1;
    }
    std::string type = argv[1];
    int subdiv = std::stoi(argv[2]);
    std::string outfile = argv[3];

    std::vector<Vector3> verts;
    std::vector<Face> faces;
    generate_mesh(type, subdiv, verts, faces);
    save_off(outfile, verts, faces);
    std::cout << "Generated " << type << " with " << verts.size() << " vertices and " << faces.size() << " triangles.\n";
    return 0;
}

File: bench/generate_mesh.py
import sys
import math

def generate_sphere(subdivisions):
    # Start with an octahedron
    f = (1.0 + math.sqrt(5.0)) / 2.0
    verts = [
        (-1,  f,  0), ( 1,  f,  0), (-1, -f,  0), ( 1, -f,  0),
        ( 0, -1,  f), ( 0,  1,  f), ( 0, -1, -f), ( 0,  1, -f),
        ( f,  0, -1), ( f,  0,  1), (-f,  0, -1), (-f,  0,  1)
    ]
    # Normalize to unit sphere
    verts = [(v[0]/math.sqrt(v[0]**2+v[1]**2+v[2]**2), 
              v[1]/math.sqrt(v[0]**2+v[1]**2+v[2]**2), 
              v[2]/math.sqrt(v[0]**2+v[1]**2+v[2]**2)) for v in verts]
    
    faces = [
        (0, 11, 5), (0, 5, 1), (0, 1, 7), (0, 7, 10), (0, 10, 11),
        (1, 5, 9), (5, 11, 4), (11, 10, 2), (10, 7, 6), (7, 1, 8),
        (3, 9, 4), (3, 4, 2), (3, 2, 6), (3, 6, 8), (3, 8, 9),
        (4, 9, 5), (2, 4, 11), (6, 2, 10), (8, 6, 7), (9, 8, 1)
    ]

    def get_middle_point(p1_idx, p2_idx, verts, middle_point_cache):
        smaller_idx = min(p1_idx, p2_idx)
        greater_idx = max(p1_idx, p2_idx)
        key = (smaller_idx, greater_idx)

        if key in middle_point_cache:
            return middle_point_cache[key]

        p1 = verts[p1_idx]
        p2 = verts[p2_idx]
        middle = ((p1[0]+p2[0])/2, (p1[1]+p2[1])/2, (p1[2]+p2[2])/2)
        # normalize
        length = math.sqrt(middle[0]**2 + middle[1]**2 + middle[2]**2)
        middle = (middle[0]/length, middle[1]/length, middle[2]/length)
        
        verts.append(middle)
        index = len(verts) - 1
        middle_point_cache[key] = index
        return index

    for _ in range(subdivisions):
        faces_subdiv = []
        cache = {}
        for tri in faces:
            v1 = tri[0]
            v2 = tri[1]
            v3 = tri[2]
            a = get_middle_point(v1, v2, verts, cache)
            b = get_middle_point(v2, v3, verts, cache)
            c = get_middle_point(v3, v1, verts, cache)
            faces_subdiv.append((v1, a, c))
            faces_subdiv.append((v2, b, a))
            faces_subdiv.append((v3, c, b))
            faces_subdiv.append((a, b, c))
        faces = faces_subdiv

    return verts, faces

def save_off(filename, verts, faces):
    with open(filename, 'w') as f:
        f.write("OFF\n")
        f.write(f"{len(verts)} {len(faces)} 0\n")
        for v in verts:
            f.write(f"{v[0]} {v[1]} {v[2]}\n")
        for tri in faces:
            f.write(f"3 {tri[0]} {tri[1]} {tri[2]}\n")

if __name__ == "__main__":
    if len(sys.argv) < 3:
        print("Usage: python3 generate_mesh.py <subdivisions> <output.off>")
        sys.exit(1)
    
    subdiv = int(sys.argv[1])
    outfile = sys.argv[2]
    
    verts, faces = generate_sphere(subdiv)
    save_off(outfile, verts, faces)
    print(f"Generated sphere with {len(verts)} vertices and {len(faces)} triangles.")
